analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.IsNull.cs

```diff
@@ -0,0 +1,208 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+using SonarAnalyzer.UnitTest.TestFramework.SymbolicExecution;
+
+namespace SonarAnalyzer.UnitTest.SymbolicExecution.Roslyn
+{
+    public partial class RoslynSymbolicExecutionTest
+    {
+        [TestMethod]
+        public void IsNull_Coalesce_SetsObjectConstraint()
+        {
+            const string code = @"
+object nullValue = null;
+object notNullValue = new object();
+var nullToNull = nullValue ?? nullValue;
+var nullToNotNull = nullValue ?? notNullValue;
+var nullToUnknown = nullValue ?? arg;
+var notNullToNull = notNullValue ?? nullValue;
+var notNullToNotNull = notNullValue ?? notNullValue;
+var notNullToUnknown = notNullValue ?? arg;
+Tag(""NullToNull"", nullToNull);
+Tag(""NullToNotNull"", nullToNotNull);
+Tag(""NullToUnknown"", nullToUnknown);
+Tag(""NotNullToNull"", notNullToNull);
+Tag(""NotNullToNotNull"", notNullToNotNull);
+Tag(""NotNullToUnknown"", notNullToUnknown);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.IsNull);
+            validator.ValidateTag("NullToNull", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("NullToNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NullToUnknown", x => x.Should().BeNull());
+            validator.ValidateTag("NotNullToNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NotNullToNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NotNullToUnknown", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+        }
+
+        [TestMethod]
+        public void IsNull_Coalesce_UnknownToNull()
+        {
+            const string code = @"
+object nullValue = null;
+var unknownToNull = arg ?? nullValue;
+Tag(""UnknownToNull"", unknownToNull);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.IsNull);
+            validator.TagValues("UnknownToNull").Should().HaveCount(2)
+                .And.ContainSingle(x => x == null)
```

Comment:
Is this the "BoolConstraint"? If so, please add a comment like `// BoolConstraint.False`

Reply:
No, it's the missing symbolic value. There's no constraint on it

Reply:
The assertion is the same as bunch of others (and yes, it would be better to have extended assertion framework for this repetitive thing with better name)